Paramiko exception: EOFError on SFTP copy

```
$ skylark deprovision; bash scripts/experiment_paras.sh aws:us-east-1 aws:us-east-2
20:54:55 [DEBUG] AWS authentication enabled, querying for instances
20:54:55 [DEBUG] Azure authentication enabled, querying for instances
20:54:56 [DEBUG] GCP authentication enabled, querying for instances
Query instances: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 24/24 [00:13<00:00,  1.72it/s]
No instances to deprovision, exiting...
+ src=aws:us-east-1
+ dest=aws:us-east-2
+ key_prefix=fake_imagenet
+ bucket_prefix=exps-paras
+ src_bucket=(${src//:/ })
+ src_bucket=exps-paras-skylark-us-east-1
+ dest_bucket=(${dest//:/ })
+ dest_bucket=exps-paras-skylark-us-east-2
+ echo exps-paras-skylark-us-east-1
exps-paras-skylark-us-east-1
+ echo exps-paras-skylark-us-east-2
exps-paras-skylark-us-east-2
+ max_instance=1
+ experiment=aws-us-east-1_aws-us-east-2_1_fake_imagenet
+ filename=data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json
+ echo data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json
data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json
+ python scripts/setup_bucket.py --key-prefix fake_imagenet --bucket-prefix exps-paras --src-data-path ../fake_imagenet/ --src-region aws:us-east-1 --dest-region aws:us-east-2
exps-paras-skylark-us-east-1
exps-paras-skylark-us-east-2
running upload... (note: may need to chunk)
  0%|                                                                                                                                                                                        | 0/1152 [00:00<?, ?it/s]waiting for completion
 77%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                                        | 882/1152 [02:02<00:31,  8.56it/s]waiting for completion
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1152/1152 [04:05<00:00,  4.69it/s]
+ source scripts/pack_docker.sh
++ BGreen='\033[1;32m'
++ NC='\033[0m'
++ echo -e '\033[1;32mBuilding docker image\033[0m'
Building docker image
++ sudo DOCKER_BUILDKIT=1 docker build -t skylark .
[+] Building 3.8s (17/17) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                             0.0s
 => => transferring dockerfile: 1.30kB                                                                                                                                                                           0.0s
 => [internal] load .dockerignore                                                                                                                                                                                0.0s
 => => transferring context: 449B                                                                                                                                                                                0.0s
 => resolve image config for docker.io/docker/dockerfile:1                                                                                                                                                       0.1s
 => CACHED docker-image://docker.io/docker/dockerfile:1@sha256:178c4e4a93795b9365dbf6cf10da8fcf517fcb4a17f1943a775c0f548e9fc2ff                                                                                  0.0s
 => [internal] load .dockerignore                                                                                                                                                                                0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                             0.0s
 => [internal] load metadata for docker.io/library/python:3.10-slim                                                                                                                                              0.1s
 => [stage-0 1/8] FROM docker.io/library/python:3.10-slim@sha256:cd62775c2141dc3d06c1b284ceda8ebe329d588ab398b0fc296248bca5fec64a                                                                                0.0s
 => [internal] load build context                                                                                                                                                                                0.0s
 => => transferring context: 871.55kB                                                                                                                                                                            0.0s
 => CACHED [stage-0 2/8] RUN (echo 'net.ipv4.ip_local_port_range = 12000 65535' >> /etc/sysctl.conf)     && (echo 'fs.file-max = 1048576' >> /etc/sysctl.conf)     && mkdir -p /etc/security/     && (echo '*    0.0s
 => CACHED [stage-0 3/8] RUN --mount=type=cache,target=/var/cache/apt apt update     && apt-get install --no-install-recommends -y git wget ca-certificates build-essential graphviz     && apt-get clean     &  0.0s
 => CACHED [stage-0 4/8] COPY scripts/requirements-gateway.txt /tmp/requirements-gateway.txt                                                                                                                     0.0s
 => CACHED [stage-0 5/8] RUN --mount=type=cache,target=/root/.cache/pip pip3 install --no-cache-dir --compile -r /tmp/requirements-gateway.txt && rm -r /tmp/requirements-gateway.txt                            0.0s
 => CACHED [stage-0 6/8] WORKDIR /pkg                                                                                                                                                                            0.0s
 => [stage-0 7/8] COPY . .                                                                                                                                                                                       0.1s
 => [stage-0 8/8] RUN pip3 install -e .                                                                                                                                                                          3.1s
 => exporting to image                                                                                                                                                                                           0.1s
 => => exporting layers                                                                                                                                                                                          0.1s
 => => writing image sha256:681e1a93070003ac800b63bf0b9a7385d080ccfa0c77d855eecc3fa78f81c2bd                                                                                                                     0.0s
 => => naming to docker.io/library/skylark                                                                                                                                                                       0.0s
+++ openssl rand -hex 16
++ DOCKER_URL=ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66
++ echo -e '\033[1;32mUploading docker image to ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66\033[0m'
Uploading docker image to ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66
++ sudo docker tag skylark ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66
++ sudo docker push ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66
The push refers to repository [ghcr.io/parasj/skylark]
31522961c895: Pushed
5117a31f2436: Pushed
3fad783f7d01: Layer already exists
52f9a593fbc0: Layer already exists
8f7843e41685: Layer already exists
5e8bd0bbd8b2: Layer already exists
a59763922a3b: Layer already exists
2267a215ad27: Layer already exists
03b10f1cb90c: Layer already exists
71e327a423e7: Layer already exists
2032d4408823: Layer already exists
3a626bb08c24: Layer already exists
local-b8b6301bbde0812e2a945c00a61bed66: digest: sha256:5fe6d938867c0fb18a6e300a4333f3297d10bbc7ec4d18e781cb4893db83f860 size: 2838
++ sudo docker system prune -f
Deleted build cache objects:
ekto5ohhmkbnraccyupyefugh
qylerd45kkx2c4lgcju8f69me
qv917d8wm1vhcf2nxnzv7uzgc

Total reclaimed space: 859.1kB
++ export SKYLARK_DOCKER_IMAGE=ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66
++ SKYLARK_DOCKER_IMAGE=ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66
++ echo -e '\033[1;32mSKYLARK_DOCKER_IMAGE=ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66\033[0m'
SKYLARK_DOCKER_IMAGE=ghcr.io/parasj/skylark:local-b8b6301bbde0812e2a945c00a61bed66
+ throughput=3
+ skylark solver solve-throughput aws:us-east-1 aws:us-east-2 3 -o data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json --max-instances 1
20:59:29 [WARN]  Gurobi not installed, using CoinOR instead.
20:59:36 [DEBUG] Solve throughput problem: 6.83s
20:59:36 [DEBUG] Total cost: $0.0108 (egress: $0.0100, instance: $0.0008)
20:59:36 [DEBUG] Total throughput: [3.0] Gbps
20:59:36 [DEBUG] Total runtime: 2.67s
20:59:36 [DEBUG] Instance regions: [aws:us-east-1=1, aws:us-east-2=1]
20:59:36 [DEBUG] Flow matrix:
20:59:36 [DEBUG] 	aws:us-east-1 -> aws:us-east-2: 3.00 Gbps with 64.0 connections, 64.0GB (link capacity = 4.59 Gbps)
/home/ubuntu/skylark/skylark/replicate/solver.py:317: RuntimeWarning: invalid value encountered in double_scalars
  average_egress_conns = [np.ceil(solution.var_conn[i, :].sum() / solution.var_instances_per_region[i]) for i in range(len(regions))]
/home/ubuntu/skylark/skylark/replicate/solver.py:318: RuntimeWarning: invalid value encountered in double_scalars
  average_ingress_conns = [np.ceil(solution.var_conn[:, i].sum() / solution.var_instances_per_region[i]) for i in range(len(regions))]
20:59:36 [WARN]  aws:us-east-1:0:0c -> aws:us-east-2: 64c of 0c remaining
20:59:36 [WARN]  aws:us-east-1:0:0c -> aws:us-east-2:0:0c: 64c remaining
20:59:36 [WARN]  Scaling connections by 1.00x
+ echo data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json
data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json
+ mkdir -p data/results
+ mkdir -p data/results/aws-us-east-1_aws-us-east-2_1_fake_imagenet
+ cp data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json data/results/aws-us-east-1_aws-us-east-2_1_fake_imagenet
+ skylark replicate-json data/plan/aws-us-east-1_aws-us-east-2_1_fake_imagenet.json --source-bucket exps-paras-skylark-us-east-1 --dest-bucket exps-paras-skylark-us-east-2 --key-prefix fake_imagenet
20:59:40 [WARN]  Deleting key skylark-us-east-1 in region us-east-1
20:59:40 [WARN]  Deleting key skylark-us-east-2 in region us-east-2
20:59:40 [INFO]  Created key file /home/ubuntu/.skylark/keys/aws/skylark-us-east-1.pem
20:59:41 [INFO]  Created key file /home/ubuntu/.skylark/keys/aws/skylark-us-east-2.pem
20:59:41 [DEBUG] Cloud SSH key initialization: 2.10s
21:00:10 [DEBUG] Provisioning instances and waiting to boot: 28.85s
21:00:11 [DEBUG] Install docker: 0.41s
21:00:11 [DEBUG] Install docker: 0.82s
21:00:24 [DEBUG] Starting gateway aws:us-east-1:i-0f7fde3b0721536f1, host: 54.236.13.116: Docker pull: 12.60s
21:00:24 [DEBUG] Starting gateway aws:us-east-1:i-0f7fde3b0721536f1, host: 54.236.13.116: Starting gateway container
21:00:25 [DEBUG] Starting gateway aws:us-east-2:i-015db4941a2a5e7e5, host: 3.135.229.95: Docker pull: 13.61s
21:00:25 [DEBUG] Starting gateway aws:us-east-2:i-015db4941a2a5e7e5, host: 3.135.229.95: Starting gateway container
21:00:26 [DEBUG] Starting gateway aws:us-east-2:i-015db4941a2a5e7e5, host: 3.135.229.95: Gateway started 39bf1d262f9b9c0a388a85bbe7a4ef60a79a1bb3ba470e416f67609d4d06dc5b
21:00:27 [DEBUG] Starting gateway aws:us-east-1:i-0f7fde3b0721536f1, host: 54.236.13.116: Gateway started 429b6fa7bb1c57f27ec4a5a7361e191205b1121efb80ff0d8bbf0643a082d23e
21:00:27 [DEBUG] Install gateway package on instances: 17.37s
21:00:27 [INFO]  Provisioned ReplicationTopologyGateway(region='aws:us-east-2', instance=0): http://3.135.229.95:8888/container/39bf1d262f9b
21:00:27 [INFO]  Provisioned ReplicationTopologyGateway(region='aws:us-east-1', instance=0): http://54.236.13.116:8888/container/429b6fa7bb1c
21:00:28 [INFO]  Batch 0 size: 64873645508 with 1002 chunks
21:00:28 [DEBUG] Building chunk requests: 0.00s
21:00:28 [DEBUG] Sending 1002 chunk requests to 54.236.13.116
21:00:28 [DEBUG] Dispatch chunk requests: 0.42s
21:00:28 [INFO]  60.42GByte replication job launched
21:00:28 [DEBUG] Registering shutdown handler
Replication: average 0.00Gbit/s:   0%|                                                                                                                                                   | 0.00/483G [10:00<?, ?bit/s]21:10:28 [ERROR] No chunks completed after 600s! There is probably a bug, check logs. Exiting...
Replication: average 0.00Gbit/s:   0%|                                                                                                                                                   | 0.00/483G [10:00<?, ?bit/s]
21:10:28 [WARN]  Deprovisioning instances
21:10:29 [WARN]  Deprovisioning aws:us-east-1:i-0f7fde3b0721536f1
21:10:29 [WARN]  Deprovisioning aws:us-east-2:i-015db4941a2a5e7e5
21:10:29 [INFO]  Copying gateway logs from aws:us-east-2:i-015db4941a2a5e7e5
21:10:29 [INFO]  Copying gateway logs from aws:us-east-1:i-0f7fde3b0721536f1
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp_client.py", line 852, in _read_response
    t, data = self._read_packet()
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp.py", line 201, in _read_packet
    x = self._read_all(4)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp.py", line 188, in _read_all
    raise EOFError()
EOFError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/skylark/skylark/utils/utils.py", line 72, in do_parallel
    args, result = future.result()
  File "/home/ubuntu/miniconda/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/home/ubuntu/miniconda/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/home/ubuntu/miniconda/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ubuntu/skylark/skylark/utils/utils.py", line 65, in wrapped_fn
    return args, func(args)
  File "/home/ubuntu/skylark/skylark/replicate/replicator_client.py", line 330, in copy_log
    instance.download_file("/tmp/gateway.stderr", transfer_dir / f"gateway_{instance.uuid()}.stderr")
  File "/home/ubuntu/skylark/skylark/compute/server.py", line 177, in download_file
    sftp.get(remote_path, local_path)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp_client.py", line 811, in get
    size = self.getfo(remotepath, fl, callback, prefetch)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp_client.py", line 786, in getfo
    return self._transfer_with_callback(
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp_client.py", line 678, in _transfer_with_callback
    data = reader.read(32768)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/file.py", line 219, in read
    new_data = self._read(read_size)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp_file.py", line 185, in _read
    t, msg = self.sftp._request(
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp_client.py", line 822, in _request
    return self._read_response(num)
  File "/home/ubuntu/miniconda/lib/python3.8/site-packages/paramiko/sftp_client.py", line 854, in _read_response
    raise SSHException("Server connection dropped: {}".format(e))
paramiko.ssh_exception.SSHException: Server connection dropped:
+ tail -1 data/results/aws-us-east-1_aws-us-east-2_1_fake_imagenet/obj-store-logs.txt
{"total_runtime_s": 54.789271, "throughput_gbits": 0.0, "monitor_status": "timed_out", "success": false}
+ echo aws-us-east-1_aws-us-east-2_1_fake_imagenet
aws-us-east-1_aws-us-east-2_1_fake_imagenet
```

--- a/skylark/compute/aws/aws_server.py
+++ b/skylark/compute/aws/aws_server.py
@@ -77,3 +77,8 @@ class AWSServer(Server):
 
     def get_ssh_cmd(self):
         return f"ssh -i {self.local_keyfile} ec2-user@{self.public_ip()}"
+
+    def get_sftp_client(self):
+        t = paramiko.Transport((self.public_ip(), 22))
+        t.connect(username="ec2-user", pkey=paramiko.RSAKey.from_private_key_file(str(self.local_keyfile)))
+        return paramiko.SFTPClient.from_transport(t)

--- a/skylark/compute/azure/azure_cloud_provider.py
+++ b/skylark/compute/azure/azure_cloud_provider.py
@@ -302,9 +302,7 @@ class AzureCloudProvider(CloudProvider):
 
     # This code, along with some code in azure_server.py, is based on
     # https://github.com/ucbrise/mage-scripts/blob/main/azure_cloud.py.
-    def provision_instance(
-        self, location: str, vm_size: str, name: Optional[str] = None, uname: str = os.environ.get("USER")
-    ) -> AzureServer:
+    def provision_instance(self, location: str, vm_size: str, name: Optional[str] = None, uname: str = "skylark") -> AzureServer:
         assert ":" not in location, "invalid colon in Azure location"
 
         if name is None:

--- a/skylark/compute/azure/azure_server.py
+++ b/skylark/compute/azure/azure_server.py
@@ -163,7 +163,7 @@ class AzureServer(Server):
         nic_poller.result()
         vnet_poller.result()
 
-    def get_ssh_client_impl(self, uname=os.environ.get("USER"), ssh_key_password="skylark"):
+    def get_ssh_client_impl(self, uname="skylark", ssh_key_password="skylark"):
         """Return paramiko client that connects to this instance."""
         ssh_client = paramiko.SSHClient()
         ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
@@ -176,3 +176,9 @@ class AzureServer(Server):
             banner_timeout=200,
         )
         return ssh_client
+
+    def get_sftp_client(self, uname="skylark", ssh_key_password="skylark"):
+        t = paramiko.Transport((self.public_ip(), 22))
+        pkey = paramiko.RSAKey.from_private_key_file(str(self.ssh_private_key), password=ssh_key_password)
+        t.connect(username=uname, pkey=pkey)
+        return paramiko.SFTPClient.from_transport(t)

--- a/skylark/compute/gcp/gcp_server.py
+++ b/skylark/compute/gcp/gcp_server.py
@@ -102,3 +102,9 @@ class GCPServer(Server):
     def get_ssh_cmd(self, uname="skylark", ssh_key_password="skylark"):
         # todo can we include the key password inline?
         return f"ssh -i {self.ssh_private_key} -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no {uname}@{self.public_ip()}"
+
+    def get_sftp_client(self, uname="skylark", ssh_key_password="skylark"):
+        t = paramiko.Transport((self.public_ip(), 22))
+        pkey = paramiko.RSAKey.from_private_key_file(str(self.ssh_private_key), password=ssh_key_password)
+        t.connect(username=uname, pkey=pkey)
+        return paramiko.SFTPClient.from_transport(t)

--- a/skylark/compute/server.py
+++ b/skylark/compute/server.py
@@ -85,6 +85,9 @@ class Server:
         else:
             self.command_log_file = None
 
+    def get_sftp_client(self):
+        raise NotImplementedError()
+
     def get_ssh_client_impl(self):
         raise NotImplementedError()
 
@@ -172,15 +175,15 @@ class Server:
 
     def download_file(self, remote_path, local_path):
         """Download a file from the server"""
-        client = self.ssh_client
-        with client.open_sftp() as sftp:
-            sftp.get(remote_path, local_path)
+        sftp_client = self.get_sftp_client()
+        sftp_client.get(remote_path, local_path)
+        sftp_client.close()
 
     def upload_file(self, local_path, remote_path):
         """Upload a file to the server"""
-        client = self.ssh_client
-        with client.open_sftp() as sftp:
-            sftp.put(local_path, remote_path)
+        sftp_client = self.get_sftp_client()
+        sftp_client.put(local_path, remote_path)
+        sftp_client.close()
 
     def copy_public_key(self, pub_key_path: PathLike):
         """Append public key to authorized_keys file on server."""

